homeassistant/components/bluecurrent/__init__.py

```diff
@@ -0,0 +1,244 @@
+            self.dispatch_signal()
+
+    async def get_charge_point_data(self, evse_id: str) -> None:
+        """Get all the data of a charge point."""
+        await self.client.get_status(evse_id)
+
+    def add_charge_point(self, evse_id: str, model: str) -> None:
+        """Add a charge point to charge_points."""
+        self.charge_points[evse_id] = {MODEL_TYPE: model}
+
+    def update_charge_point(self, evse_id: str, data: dict) -> None:
+        """Update the charge point data."""
+        for key in data:
+            self.charge_points[evse_id][key] = data[key]
+        self.dispatch_signal(evse_id)
+
+    def dispatch_signal(self, evse_id: str | None = None) -> None:
+        """Dispatch a signal."""
+        if evse_id is not None:
+            async_dispatcher_send(self._hass, f"{DOMAIN}_value_update_{evse_id}")
+        else:
+            async_dispatcher_send(self._hass, f"{DOMAIN}_grid_update")
+
+    async def start_loop(self) -> None:
+        """Start the receive loop."""
+        try:
+            await self.client.start_loop(self.on_data)
+        except (BlueCurrentException) as err:
+            LOGGER.warning(
+                "Disconnected from the Blue Current websocket. Retrying to connect in background. %s",
+                err,
+            )
+
+            async_call_later(self._hass, DELAY_1, self.reconnect)
+
+    async def reconnect(self, event_time: datetime | None = None) -> None:
+        """Keep trying to reconnect to the websocket."""
+        try:
+            await self.connect(self._config.data[CONF_API_TOKEN])
+            LOGGER.warning("Reconnected to the Blue Current websocket")
+            self._hass.loop.create_task(self.start_loop())
+            await self.client.get_charge_points()
+        except (WebsocketException, RequestLimitReached) as err:
+            set_entities_unavalible(self._hass, self._config.entry_id)
+            if isinstance(err, RequestLimitReached):
+                delay = self.client.get_next_reset_delta()
+            else:
+                delay = DELAY_2
+            async_call_later(self._hass, delay, self.reconnect)
+
+    async def disconnect(self) -> None:
+        """Disconnect from the websocket."""
+        try:
+            await self.client.disconnect()
+        except WebsocketException:
+            pass
+
+
+class BlueCurrentEntity(Entity):
+    """Define a base charge point entity."""
+
+    def __init__(self, connector: Connector, evse_id: str | None = None) -> None:
+        """Initialize the entity."""
+        self._connector: Connector = connector
+
+        if evse_id:
+            self._evse_id = evse_id
+            self._attr_device_info = DeviceInfo(
+                identifiers={(DOMAIN, evse_id)},
+                name=evse_id,
+                manufacturer="BlueCurrent",
+                model=connector.charge_points[evse_id][MODEL_TYPE],
+            )
+            self.is_grid = False
+        else:
+            self.is_grid = True
```

Comment:
This sounds like you'd really want two classes, a ChargePoint entity and a Grid entity, instead of one that can act as two depending on the flag?

Reply:
Split the sensors into classes.